### ISSUE TITLE
[tests,pmp] (Potential) bug with PMP smoketest.

### ISSUE THREAD

**Issue by timothytrippel** [lowRISC/opentitan#9508]
# Overview

When migrating all chip-level tests to the OTTF (in #9368), I encountered test failures for both the PMP smoke tests:
- `pmp_smoketest_napot`
- `pmp_smoketest_tor`

# (Potential) Bug Description

After a quick debugging session, I noticed some strange behavior. Specifically:

1. [this line](https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/tests/pmp_smoketest_tor.c#L115) first triggers an exception, as it should since it's testing the PMP causes this behavior, then
2. the OTTF saves the execution context on the stack ([including the `mepc`](https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/lib/testing/test_framework/ottf_isrs.S#L118), since this is required for a test to run as a FreeRTOS task if the  [`enable_concurrency`](https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/lib/testing/test_framework/ottf.h#L44) flag is set in the test configuration struct), then
3. the OTTF correctly handles this exception, and sets a boolean flag indicating to the test that an exception occurred as expected, then
4. the OTTF restores the `mepc` from the stack before calling `mret` (though it should not have changed since the exception handler was entered since the there is no context switch happening in these bare-metal tests), after which
5. a different value is loaded into the PC shortly after the `mepc` is loaded back into the `pc` from the `mret` being issued, which causes an immediate illegal instruction exception to be thrown (since this strange`pc` value seems to be not aligned to the instruction boundary, almost as if an offset from a compressed instruction was computed) 

To show this behavior, I have included both the GDB trace (below) I captured where I set a single breakpoint at the `mret` instruction issued by the OTTF ISR exit subroutine.

<img width="726" alt="Screen Shot 2021-12-02 at 5 36 24 PM" src="https://user-images.githubusercontent.com/5633066/144532813-5f274090-1530-4045-bed0-4e682719faa5.png">

Additionally, I captured a waveform (below), and highlight the area where this strange behavior has occurred.

<img width="2209" alt="Screen Shot 2021-12-02 at 6 30 19 PM" src="https://user-images.githubusercontent.com/5633066/144534766-caf96f44-08b7-4dac-91d4-413f8a7e60f9.png">

What is more puzzling is that the first PMP exception-raise-check on [this line](https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/tests/pmp_smoketest_tor.c#L110) (which checks that an attempted load of the first index of a protected array raises an exception) seems to work fine, but it is the second check on [this line](https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/tests/pmp_smoketest_tor.c#L115) (which checks that an attempted load of the last index of a protected array raises an exception) causes the behavior described (and shown in the GDB trace) above.

# Temporary fix to unblock use of the OTTF

Additionally, it seems that reducing size of the [PMP guarded test array ](https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/tests/pmp_smoketest_tor.c#L46) from 2048 bytes to 1024 bytes makes the (potential) bug disappear. Which is why in #9368, I made [this change](https://github.com/lowRISC/opentitan/pull/9368/commits/3df509d0b049a190b0ed1de34cf703b12326568d#r761609098) to the PMP smoke tests so that they would pass CI so I could un-block migrating the tests to the OTTF. However I wanted to document this (potential) issue just in case this is of concern. I say _potential_ because it seems as though this may be some memory corruption bug caused by SW configs, than HW itself, but it is unclear to me at this time, as I have yet to be able to dig into this further.

# Reproducing this Issue

I have created a [branch](https://github.com/timothytrippel/opentitan/tree/pmp-smoketest-bug) in my forked OT repo with the old versions of the PMP smoke tests (prior to their change in #9368) that can be used to reproduce this bug.

Thoughts?

**Comment by GregAC** [comment]
Could you provide me with the .fst file you got the waveform screenshot from? I can do some digging around and try to confirm whether or not we have a hardware issue.

You should just be able to attach it to this issue, if it's too large upload to the OT gdrive and email me the link.

**Comment by mundaym** [comment]
Both OTTF and the test are attempting to skip to the next instruction by updating `mepc`. Is there some sort of interaction going on there? It looks like OTTF should 'win` and skip just one instruction but perhaps I'm missing something.

By the way I'm not sure the OTTF should really be updating `mepc` in the way described here:

https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/lib/testing/test_framework/ottf_isrs.S#L20-L22

I think that the operating system should allow the *application* to decide what the `mepc` value should be, even if it means that the application could enter an infinite loop. Skipping the instruction means that the operating system is trying to handle an exception for the application and that could result in some undesirable behaviour. If the application has registered its own exception handler then the operating system should let that exception handler set `mepc`.

**Comment by timothytrippel** [comment]
> Both OTTF and the test are attempting to skip to the next instruction by updating `mepc`. Is there some sort of interaction going on there? It looks like OTTF should 'win` and skip just one instruction but perhaps I'm missing something.

My branch I linked above contains an updated version of the PMP smoke test  that runs on the OTTF [that does **_not_** update the `mepc` itself](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/tests/pmp_smoketest_tor.c#L48) in the overridden OTTF exception handler, rather it leaves that to the OTTF exception [entry](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.S#L115-L118) / [exit](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.S#L341-L343) asm subroutines that run before / after the overridden exception handler that is written in the test itself (which just sets a global flag indicating the exception occurred). So only in one place is the `mepc` updated to skip to the next instruction. Does that make sense?

> By the way I'm not sure the OTTF should really be updating `mepc` in the way described here:
> 
> https://github.com/lowRISC/opentitan/blob/caeb0b7ea2966dadbaa60357e045ff64d68e8c72/sw/device/lib/testing/test_framework/ottf_isrs.S#L20-L22
> 
> I think that the operating system should allow the _application_ to decide what the `mepc` value should be, even if it means that the application could enter an infinite loop. Skipping the instruction means that the operating system is trying to handle an exception for the application and that could result in some undesirable behaviour. If the application has registered its own exception handler then the operating system should let that exception handler set `mepc`.

First, I want to clarify how the OTTF handles all interrupt/exceptions.

For all **_exceptions_** the following subroutines/functions are implemented (note the weak symbols can be overridden by the application, i.e. test):
(strong symbol / asm) --> (weak symbol / C) --> (weak symbol / C) --> (strong symbol / asm)
[`handler_exception`](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.S#L79) --> [`ottf_exception_handler`](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.c#L13) --> cause-specific handler, e.g. [`ottf_load_store_fault_handler`](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.c#L71) --> [`ottf_isr_exit`](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.S#L327)

For all **_interrupts_** the follow subroutines/functions are implemented (again note which symbols are weak):
(strong symbol / asm) --> (weak symbol / C) --> (strong symbol / asm)
[`handler_irq_<software | timer | external>`](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.S#L141-L317) --> [`ottf_<software | timer | external>_isr `](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.c#L87-L100) --> [`ottf_isr_exit`](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/lib/testing/test_framework/ottf_isrs.S#L327)

The reason it was written this way was because FreeRTOS uses `ecalls` to support non-preemptive context switching upon a task invoking `yield`, and it uses timer interrupts to support preemptive context switching. If an exception (or timer ISR) is invoked by a call to `yield` (or the timer tick), and a context switch is about to happen, then the current task's `mepc` must be saved to the stack  before the context switch, as another task's `mepc` is will overwrite the `mepc` in the `ottf_isr_exit` subroutine. By updating the `mepc` to the next instruction (for _exceptions_ and _software IRQs_) in the strong symbols (`handler_exception` and `handler_irq_software`) vs. any of the override-able weak symbol, it alleviates the burden on test developers to do this step (this was in fact how the FreeRTOS-provided RISC-V port was written). If a test developer really wants to alter the `mepc` themselves in their test application, they can still do so by overriding the (weak) parent exception handler (`ottf_exception_handler`) or software ISR `ottf_software_isr` and overwriting the location on the stack where the `mepc` is stored by the (strong / asm) `handler_exception` and `handler_irq_software` symbols. This was the approach I took to [modify the `sram_ctrl_execution_test_ret.c` test](https://github.com/timothytrippel/opentitan/blob/5fff7f20eaac3aaa3f881894b7d994ad081a6a65/sw/device/tests/sram_ctrl_execution_test_ret.c#L58-L96) that @silvestrst originally wrote to migrate it to the OTTF in #9368.

Does this make sense? More than happy to move the `mepc` update (for synchronous irqs/exceptions) to the weak symbols, or provide OTTF APIs for modifying the `mepc` in a overridden exception handler / ISR, if you think this is a better approach. Want to make sure I get this right / make life as easy as possible for test writers. Lmk what you think!

**Comment by tjaychen** [comment]
is this still a bug @timothytrippel ? looking back i wonder if this could be related to the same issue @cfrantz ran into regarding memory initialization.

**Comment by alphan** [comment]
friendly ping @timothytrippel
